mobile/library/java/io/envoyproxy/envoymobile/engine/AndroidNetworkMonitorV2.java

```diff
@@ -0,0 +1,809 @@
+           // so networkInfo.getType() can't be matched against
+           // defaultNetworkInfo.getType() but networkInfo.getType() should
+           // be TYPE_VPN. In the case of a VPN, getAllNetworks() will have
+           // returned just this VPN if it applies.
+           || networkInfo.getType() == TYPE_VPN)) {
+        // Android 10+ devices occasionally return multiple networks
+        // of the same type that are stuck in the CONNECTING state.
+        // Now that Java asserts are enabled, ignore these zombie
+        // networks here to avoid hitting the assert below. crbug.com/1361170
+        if (defaultNetwork != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+          // If `network` is CONNECTING, ignore it.
+          if (networkInfo.getDetailedState() == NetworkInfo.DetailedState.CONNECTING) {
+            continue;
+          }
+          // If `defaultNetwork` is CONNECTING, ignore it.
+          NetworkInfo prevDefaultNetworkInfo = getRawNetworkInfo(defaultNetwork);
+          if (prevDefaultNetworkInfo != null &&
+              prevDefaultNetworkInfo.getDetailedState() == NetworkInfo.DetailedState.CONNECTING) {
+            defaultNetwork = null;
+          }
+        }
+        if (defaultNetwork != null) {
+          // TODO(crbug.com/40060873): Investigate why there are multiple
+          // connected networks.
+          Log.e(TAG, "There should not be multiple connected "
+                         + "networks of the same type. At least as of Android "
+                         + "Marshmallow this is not supported. If this becomes "
+                         + "supported this assertion may trigger.");
+        }
+        defaultNetwork = network;
+      }
+    }
+    return defaultNetwork;
+  }
+
+  /**
+   * @param networkInfo The NetworkInfo for the active network.
+   * @return the info of the network that is available to this app.
+   */
+  private NetworkInfo processActiveNetworkInfo(NetworkInfo networkInfo) {
+    if (networkInfo == null) {
+      return null;
+    }
+
+    if (networkInfo.isConnected()) {
+      return networkInfo;
+    }
+
+    if (networkInfo.getDetailedState() != NetworkInfo.DetailedState.BLOCKED) {
+      // Network state is not blocked which implies that network access is
+      // unavailable (not just blocked to this app).
+      return null;
+    }
+
+    // If |networkInfo| is BLOCKED, but the app is in the foreground, then it's likely that
+    // Android hasn't finished updating the network access permissions as BLOCKED is only
+    // meant for apps in the background. See https://crbug.com/677365 for more details.
+    // TODO(danzh) check whether application is in the foreground or not.
+    return null;
+    /*
+    // fork
+    https://source.chromium.org/chromium/chromium/src/+/main:base/android/java/src/org/chromium/base/ApplicationStatus.java
+    if (ApplicationStatus.getStateForApplication()
```

Comment:
does this not work?  it seems to work for Chromium: https://source.chromium.org/chromium/chromium/src/+/main:net/android/java/src/org/chromium/net/NetworkChangeNotifierAutoDetect.java;l=233-237

Reply:
We don't have similar [ApplicationStatus]( https://source.chromium.org/chromium/chromium/src/+/main:base/android/java/src/org/chromium/base/ApplicationStatus.java;drc=7c2b25f6a19cfeeea67f0f43ed33617840bab33d;l=49) implemented in Envoy yet.